File: src/main/services/fileCache.ts
import { readdirSync, statSync, existsSync, readFileSync } from 'fs'
import { join, relative } from 'path'
import chokidar from 'chokidar'
import type { FSWatcher } from 'chokidar'

const EXCLUDED_DIRS = new Set([
  '.git',
  '.svn',
  '.hg',
  'node_modules',
  'dist',
  'build',
  'out',
  '.next',
  '.nuxt',
  '.svelte-kit',
  '__pycache__',
  '.venv',
  'venv',
  '.cache',
  '.parcel-cache',
  '.turbo'
])

const EXCLUDED_FILES = new Set([
  '.DS_Store',
  'Thumbs.db',
  'package-lock.json',
  'yarn.lock',
  'pnpm-lock.yaml'
])

export interface FileEntry {
  path: string
  type: 'file' | 'directory'
}

class FileCacheService {
  private caches: Map<string, FileEntry[]> = new Map()
  private watchers: Map<string, FSWatcher> = new Map()

  async buildCache(directory: string): Promise<FileEntry[]> {
    if (this.caches.has(directory)) {
      return this.caches.get(directory)!
    }

    const files = await this.scanDirectory(directory)
    this.caches.set(directory, files)
    this.startWatcher(directory)
    return files
  }

  getCache(directory: string): FileEntry[] | undefined {
    return this.caches.get(directory)
  }

  private async scanDirectory(directory: string, baseDir?: string): Promise<FileEntry[]> {
    const results: FileEntry[] = []
    const base = baseDir ?? directory

    if (!existsSync(directory)) {
      return results
    }

    try {
      const entries = readdirSync(directory)

      for (const entry of entries) {
        if (EXCLUDED_FILES.has(entry)) continue

        const fullPath = join(directory, entry)
        const relativePath = relative(base, fullPath)

        try {
          const stat = statSync(fullPath)
          if (stat.isDirectory()) {
            if (EXCLUDED_DIRS.has(entry)) continue
            results.push({ path: relativePath, type: 'directory' })
            const subFiles = await this.scanDirectory(fullPath, base)
            results.push(...subFiles)
          } else if (stat.isFile()) {
            results.push({ path: relativePath, type: 'file' })
          }
        } catch {
          // Skip inaccessible files
        }
      }
    } catch {
      // Skip inaccessible directories
    }

    return results
  }

  private startWatcher(directory: string): void {
    if (this.watchers.has(directory)) return

    const watcher = chokidar.watch(directory, {
      ignored: [/(^|[/\\])\./, /node_modules/, /dist/, /build/, /\.git/, /\.cache/],
      persistent: true,
      ignoreInitial: true
    })

    watcher.on('add', () => this.invalidateCache(directory))
    watcher.on('unlink', () => this.invalidateCache(directory))
    watcher.on('addDir', () => this.invalidateCache(directory))
    watcher.on('unlinkDir', () => this.invalidateCache(directory))

    this.watchers.set(directory, watcher)
  }

  private invalidateCache(directory: string): void {
    this.caches.delete(directory)
    this.buildCache(directory)
  }

  stopWatching(directory: string): void {
    const watcher = this.watchers.get(directory)
    if (watcher) {
      watcher.close()
      this.watchers.delete(directory)
      this.caches.delete(directory)
    }
  }

  async parseMentions(directory: string, spec: string): Promise<string> {
    const mentionRegex = /@([^\s]+)/g
    const files = await this.buildCache(directory)
    const fileSet = new Map(files.map((f) => [f.path, f]))

    let result = spec
    let match
    const processedPaths = new Set<string>()

    while ((match = mentionRegex.exec(spec)) !== null) {
      const mentionPath = match[1]
      if (processedPaths.has(mentionPath)) continue
      processedPaths.add(mentionPath)

      const file = fileSet.get(mentionPath)
      if (file && file.type === 'file') {
        try {
          const content = readFileSync(join(directory, mentionPath), 'utf-8')
          const insertion = `\n@${mentionPath}\n---\n${content.slice(0, 100000)}\n---\n`
          result = result.replace(match[0], insertion)
        } catch {
          // Skip files that can't be read
        }
      }
    }

    return result
  }
}

export const fileCacheService = new FileCacheService()

File: src-tauri/src/file_cache.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use tauri::Manager;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileEntry {
    pub path: String,
    #[serde(rename = "type")]
    pub file_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileListOptions {
    pub pair_id: Option<String>,
    pub directory: Option<String>,
}

const EXCLUDED_DIRS: &[&str] = &[
    ".git", ".svn", ".hg", "node_modules", "dist", "build", "out",
    ".next", ".nuxt", ".svelte-kit", "__pycache__", ".venv", "venv",
    ".cache", ".parcel-cache", ".turbo", ".pair", ".opencode"
];

const EXCLUDED_FILES: &[&str] = &[
    ".DS_Store", "Thumbs.db", "package-lock.json", "yarn.lock",
    "pnpm-lock.yaml", "*.log"
];

pub fn should_exclude_dir(name: &str) -> bool {
    EXCLUDED_DIRS.contains(&name) || name.starts_with('.')
}

pub fn should_exclude_file(name: &str) -> bool {
    EXCLUDED_FILES.contains(&name) || name.ends_with(".log")
}

pub fn scan_directory(directory: &Path, base_dir: &Path) -> Result<Vec<FileEntry>, String> {
    let mut results = Vec::new();
    
    if !directory.exists() {
        return Ok(results);
    }

    let entries = fs::read_dir(directory)
        .map_err(|e| format!("Failed to read directory: {}", e))?;

    for entry in entries {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };

        let file_name = match entry.file_name().to_str() {
            Some(name) => name.to_string(),
            None => continue,
        };

        if should_exclude_file(&file_name) {
            continue;
        }

        let full_path = entry.path();
        let relative_path = match full_path.strip_prefix(base_dir) {
            Ok(p) => p.to_string_lossy().to_string(),
            Err(_) => continue,
        };

        if full_path.is_dir() {
            if should_exclude_dir(&file_name) {
                continue;
            }
            
            results.push(FileEntry {
                path: relative_path.clone(),
                file_type: "directory".to_string(),
            });

            let sub_files = scan_directory(&full_path, base_dir)?;
            results.extend(sub_files);
        } else if full_path.is_file() {
            results.push(FileEntry {
                path: relative_path,
                file_type: "file".to_string(),
            });
        }
    }

    Ok(results)
}

#[tauri::command]
pub async fn file_list_files(
    app: tauri::AppHandle,
    options: FileListOptions,
) -> Result<Vec<FileEntry>, String> {
    let directory = if let Some(pair_id) = options.pair_id {
        let pair_manager = app.state::<std::sync::Mutex<crate::pair_manager::PairManager>>();
        let manager = pair_manager.lock().map_err(|e| e.to_string())?;
        let pairs = manager.list_pairs();
        pairs.iter()
            .find(|p| p.pair_id == pair_id)
            .map(|p| p.directory.clone())
            .ok_or_else(|| format!("Pair {} not found", pair_id))?
    } else if let Some(dir) = options.directory {
        dir
    } else {
        return Err("Either pair_id or directory must be provided".to_string());
    };

    let dir_path = Path::new(&directory);
    if !dir_path.exists() {
        return Err(format!("Directory does not exist: {}", directory));
    }

    scan_directory(dir_path, dir_path)
}

#[tauri::command]
pub async fn file_parse_mentions(
    _app: tauri::AppHandle,
    _pair_id: String,
    spec: String,
) -> Result<String, String> {
    // For now, return the spec as-is
    // TODO: Implement mention parsing to inject file contents
    Ok(spec)
}
